### PR TITLE
Add dot notation accessing for graphql responses

### LINKED ISSUE #1210: GraphQL Client inadequate since v10 upgrade
# Overview
[graphql-client](https://github.com/github/graphql-client) gem dependency was removed in the [shopify-api-ruby v10 upgrade](https://github.com/Shopify/shopify-api-ruby#breaking-change-notice-for-version-1000).

Instead of being able to access methods from objects to retrieve GraphQL response data, we return an HTTPResponse object instead, which means devs would have to retrieve data via hashes...

Before:
```
client = ShopifyAPI::GraphQL.client

SHOP_NAME_QUERY = client.parse <<-'GRAPHQL'
  {
    shop {
      name
    }
  }
GRAPHQL

result = client.query(SHOP_NAME_QUERY)
shop_name = result.data.shop.name
```

After:
```
client = ShopifyAPI::Clients::Graphql::Admin.new(session: session)

# make the GraphQL query string
query =<<~QUERY
  {
    shop {
      name
    }
  }
QUERY

response = client.query(query: query)
shop_name = response.body["data"]["shop"]["name"]
```

As you can see, this make migration and usage of the response data very difficult..
`shop_name = result.data.shop.name` vs 
`shop_name = response.body["data"]["shop"]["name"]`
It's especially challenging if there are even more layers like

`result.data.webhookSubscriptions.edges[0].node.id` vs
`response.body["data"]["webhookSubscriptions"]["edges"][0]["node"]["id"]`

## Expected behavior
We should add support to allow method access for GraphQL data instead of hash access to avoid this undesirable v10 migration side effect for existing apps, and improve new developer's GraphQL experience when building new apps.

[Link to Comment](https://github.com/Shopify/shopify-api-ruby/issues/1032#issuecomment-1279864265)

### PR DESCRIPTION
**Pull request by lizkenyon**
## Description

https://github.com/Shopify/shopify-api-ruby/assets/23265671/144cffa0-aa34-4a5b-a51a-38c07299b5cd

Fixes #1210

Please, include a summary of what the PR is for:
Starting in the version 10 upgrade graphql response were no longer available to be access with dot notation. 

This PR adds a context option(`response_as_struct`) that will return the  graphql API responses as a struct, so that they will be accessible via either dot or hash notation.

```ruby
 created_product = response.body["data"]["productCreate"]["product"]
 created_product2 = response.body.data.productCreate.product
```

## How has this been tested?

* This gem was used locally to test GraphQL, REST, rest resources and install flow

## Checklist:

- [ ] My commit message follow the pattern described in [here](https://chris.beams.io/posts/git-commit/)
- [ ] I have performed a self-review of my own code.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have updated the project documentation.
- [ ] I have added a changelog line.

### CODE REVIEW

File: lib/shopify_api/context.rb [hunk 1]
```diff
@@ -128,6 +132,9 @@ def load_rest_resources(api_version:)
       sig { returns(Symbol) }
       attr_reader :log_level
 
+      sig { returns T.nilable(T::Boolean) }
+      attr_reader :graphql_response_object
```
Comment (lizkenyon) [inline lib/shopify_api/context.rb:136]: When do we use attr_reader vs manually creating our own accessor methods?

File: test/clients/http_response_test.rb [hunk 2]
```diff
@@ -45,6 +45,15 @@ def test_next_and_prev
         assert_equal("page-info", response.prev_page_info)
         assert_equal("other-page-info", response.next_page_info)
       end
+
+      def test_object_response
```
Comment (lizkenyon) [inline test/clients/http_response_test.rb:49]: Should we duplicate all the tests in this file with the context_object set to true, to test we are not breaking the original functionality?

File: lib/shopify_api/clients/http_client.rb [hunk 3]
```diff
@@ -32,8 +32,8 @@ def initialize(base_path:, session: nil)
         end
       end
 
-      sig { params(request: HttpRequest).returns(HttpResponse) }
-      def request(request)
+      sig { params(request: HttpRequest, graphql_response_object: T::Boolean).returns(HttpResponse) }
+      def request(request, graphql_response_object: false)
```
Comment (nelsonwittwer) [inline lib/shopify_api/clients/http_client.rb:36]: This is an even safer approach, I like it!

Maybe we rename this param as `response_as_struct: false` to be a little more explicit as what the option is doing.

**Comment by oliwoodsuk** [comment]
Hey guys,
Adding this option throws a type error from this [file](https://github.com/Shopify/shopify-api-ruby/blob/main/lib/shopify_api/webhooks/registry.rb) when the option is setup in `ShopifyAPI::Context` like so:
```ruby
ShopifyAPI::Context.setup(
            ...
            response_as_struct: true
        )
```

Thanks for the change though. We're currently migrating from an older version, but are just going to re-instantiate the response using `ActiveSupport::OrderedOptions` for now to get dot notation.

**Comment by oliwoodsuk** [comment]
Hey @lizkenyon ,

Yep sure, we're not using the ruby template. Instead we're using the `shopify_app` gem @ `21.8.0` and the `shopify-api-ruby` gem effectively on main (through our own fork), only changes are the addition of a `lib/shopify_api/rest/resources/2024_01` folder and `"2024-01"` to `ShopifyAPI::AdminVersions::SUPPORTED_ADMIN_VERSIONS`

We're running a SPA style app via Turbo/Hotwire and below is the `initializers/shopify_app.rb` file:

```ruby
API_DOMAIN     = (Rails.env.production?) ? 'https://api.xxx.io'                   : 'https://xxx.ngrok.io'
WEBHOOK_DOMAIN = (Rails.env.production?) ? 'https://xxx-new-webhook.onrender.com' : 'https://xxx.ngrok.io'
APP_DOMAIN     = (Rails.env.production?) ? 'https://xxx-app.onrender.com'         : 'https://xxx.ngrok.io'

api_scopes =  "read_products,write_products,read_script_tags,write_script_tags,read_themes,write_themes,read_orders,write_orders,read_inventory,write_inventory,read_price_rules,write_price_rules,read_discounts,write_discounts,read_customers,write_customers,read_order_edits,write_order_edits,read_merchant_managed_fulfillment_orders,write_merchant_managed_fulfillment_orders,read_assigned_fulfillment_orders,write_assigned_fulfillment_orders,read_third_party_fulfillment_orders,write_third_party_fulfillment_orders,read_purchase_options,write_purchase_options,read_payment_mandate,write_payment_mandate"
api_scopes += ",read_all_orders" if Rails.env.production? #requires permission from shopify to request this scope, which we don't have for dev app


ShopifyApp.configure do |config|
    config.application_name = "xxx"
    config.old_secret = ""
    config.embedded_app = true
    config.after_authenticate_job = false
    config.api_version = '2024-01'
    config.shop_session_repository = 'Shop'
    config.scope =  api_scopes
    config.log_level = :info

    config.api_key = ENV['SHOPIFY_API_KEY'],
    config.secret  = ENV['SHOPIFY_API_SECRET']

    config.scripttags = [
        {event:'onload', src:"#{ API_DOMAIN }/xxx-embed.js"}
    ]
    config.webhooks = [
        {topic: 'orders/create',   address:  "#{ WEBHOOK_DOMAIN }/webhooks/orders_create",   path: 'webhooks/orders_create',   format: 'json'},
        {topic: 'orders/updated',  address:  "#{ WEBHOOK_DOMAIN }/webhooks/orders_updated",  path: 'webhooks/orders_updated',  format: 'json'},
        {topic: 'products/update', address:  "#{ WEBHOOK_DOMAIN }/webhooks/products_update", path: 'webhooks/products_update', format: 'json'},
        {topic: 'products/delete', address:  "#{ WEBHOOK_DOMAIN }/webhooks/products_delete", path: 'webhooks/products_delete', format: 'json'},
        {topic: 'app/uninstalled', address:  "#{ WEBHOOK_DOMAIN }/webhooks/app_uninstalled", path: 'webhooks/app_uninstalled', format: 'json'}
    ]

    if defined? Rails::Server
        raise('Missing SHOPIFY_API_KEY. See https://github.com/Shopify/shopify_app#requirements') unless config.api_key
        raise('Missing SHOPIFY_API_SECRET. See https://github.com/Shopify/shopify_app#requirements') unless config.secret
    end
end

Rails.application.config.after_initialize do
    if ShopifyApp.configuration.api_key.present? && ShopifyApp.configuration.secret.present?
        ShopifyAPI::Context.setup(
            api_key:            ENV['SHOPIFY_API_KEY'],
            api_secret_key:     ENV['SHOPIFY_API_SECRET'],
            host:               APP_DOMAIN,
            scope:              api_scopes,
            is_embedded:        true,
            is_private:         false,
            api_version:        ShopifyApp.configuration.api_version,
            user_agent_prefix:  "xxx",
            log_level:          Rails.env.test? ? :error : :info,
            response_as_struct: true #currently running this as false to avoid mentioned error
        )
        
        ShopifyApp::WebhooksManager.add_registrations
    end
end
```

**Comment by oliwoodsuk** [comment]
Hey @lizkenyon , maybe it was just an odd side effect from that trailing comma turning the `.api_key` method response into an array. Thanks for looking into it!

**Pull request merged**